### DISCUSSION TITLE
Implement closures for non-amd64 platforms

### DISCUSSION THREAD

**Discussion by maddanio** [vlang/v#11917] [Ideas]
Is there any idea for how to go beyond amd64? I can put in some work, because I would like to play with v, and need closures to do anything interesting :). Currently starting to look into libjit, not sure where that will lead.

**Comment by medvednikov** [comment]
What's your platform? arm64?

**Comment by maddanio** [comment]
yes, ok, libjit is gpl, but reading this now: https://hokstad.com/how-to-implement-closures

**Comment by Gladear** [comment]
The [current implementation](https://github.com/vlang/v/blob/master/vlib/v/gen/c/cheaders.v#L58) should be flexible enough to support all architectures. From what I read in your article, the author uses a similar idea, using dynamically generated assembly to push an new argument.

**Comment by maddanio** [comment]
that current implementation is amd64 specific, observe __closure_thunk, thats encoded amd64 assembly

**Comment by maddanio** [comment]
> What's your platform? arm64?

in case that got lost: yes, running on an arm64 m1 macbook pro. arm will also be important for ios and android support, which you both are at least considering, right?

**Comment by Gladear** [comment]
Indeed, I did it, I (approximately) know what I did 😄
What I say is that the only modification required should be to change the assembly

**Comment by maddanio** [comment]
well, "only" is a big roadnblock at least for me. would like to drive this forward, but i dont see myself hacking arm64 assembly :-D

**Comment by Gladear** [comment]
> which you both are at least considering, right?

The goal is to support all architectures C compilers support, it's just not done right now

**Comment by maddanio** [comment]
the linked blog article also seems to suggest that this solution is very amd64 specific? thus i started looking elsewhere

**Comment by maddanio** [comment]
so in order to drive this forward I will need to figure out the required bit of "magic" assembly? if thats the concensus then I will try to move forward there. always something new to learn :)

**Comment by Gladear** [comment]
Yes, that's the idea. The assembly should be pretty straightforward:
- push a new argument
- call the original function

And repeat for each architecture 😄

**Comment by maddanio** [comment]
so a regular function call preamble? with stack handling and all?

**Comment by maddanio** [comment]
something like documented here:
https://developer.arm.com/documentation/ihi0055/latest
?

**Comment by maddanio** [comment]
btw, can a closure return anything in v?

**Comment by Gladear** [comment]
Not exactly a regular function call. In amd64, the asm basically says
- put the context in the next registry (aka put the context in the next argument)
- call the original function

It would be different in amd86 because arguments are put on the stack, so you'd have to push all the arguments given to the stack, plus the context.
I don't know how arguments are handled in arm64, and I'm afrais it's not currently my priority. I'm trying to get closures to work on Windows, but it's reeeally complicated.... Hopefully, after that I'll fix some bugs in current implementation for amd64, and then I can look at something else (not sure when that will be...)

**Comment by maddanio** [comment]
hmm, well, since I would really like to be able to properlky use v on my m1 i will see what I can do

**Comment by maddanio** [comment]
is there a test suite for closures so I can test wether I succeeded?

**Comment by Gladear** [comment]
Sure !
`vlib/v/tests/closure_test.v`

It isn't exhaustive, and there are many more tests to add, but it's a start 🙂
